Support all math rendering engines from Pandoc in HTML documents

atusy opened the pull request:
This PR adds 

- support for all the math rendering engines supported by Pandoc even if self contained.

```yaml
output:
  html_document:
    math: katex
```

```yaml
output:
  html_document:
    math:
      katex: https://cdn.jsdelivr.net/npm/katex@0.12.0/dist
```


- ~support for user templates~
    - ~note that template must avoid `$if(math)$$math$$endif$`~
    - user templates need `self_contained: FALSE` following the current implementation.

<!-- -->

- support for local copy of MathJax by
	- Download and extract https://github.com/mathjax/MathJax/releases/tag/2.7.9
	- Specify js file via yaml front matter

```yaml
output:
  html_document:
    mathjax: MathJax-2.7.9/MathJax.js?config=TeX-AMS-MML_HTMLorMML
```

atusy commented:
It seems like math rendering fails when runtime is shiny...

RLesur commented:
> note that template must avoid `$if(math)$$math$$endif$`

IMO, this is an antipattern choice. 
For instance, the Pandoc's default HTML templates use `$if(math)$$math$$endif$` (see [here](https://github.com/jgm/pandoc-templates/blob/f11a6ef3c04e0f7b95179ca768c07aa9657ac196/default.html5#L23-L25) and [here](https://github.com/jgm/pandoc-templates/blob/f11a6ef3c04e0f7b95179ca768c07aa9657ac196/default.html4#L23-L25)). It would be awkward that `rmarkdown::html_document()` would be incompatible with Pandoc's default templates.

atusy commented:
@RLesur 
I see...
Currently, `html_document()` ignores the `mathjax` argument when `template` other than "default" is given.
Maybe I should do the same way.

py9mrg commented:
Great news! Thanks for doing this. I just happened to find [liminal](https://github.com/jonathanlilly/liminal) yesterday - wonder if it's any use to get around the issue mentioned in the comments? Maybe completely irrelevant though, my knowledge of all this is almost zero at the moment.

atusy commented:
I dropped the math rendering engine support for user templates as suggested by https://github.com/rstudio/rmarkdown/pull/1940#issuecomment-719921544.
Also, most of the rendering engines now work with `runtime: shiny`.
I do not understand why, but KaTeX does not work with `runtime: shiny`...

@RLesur 
I fixed the antipattern.
Apart from that, I moved MathJax to `<head>` and enabled it for `self_contained` document without a JavaScript trick.
However, we can keep the previous implementation just to make sure the consistent behavior from the older versions.
What do you think?
https://github.com/rstudio/rmarkdown/pull/1940/files#diff-91b59569aab1b937e1e5a6e1559a62de19e04f4c4bad52bfb08b910281e0f2dfR484-R486

@py9mrg 
Your welcome. Remark.js uses its own markdown parser, and thus is the different story. Thanks, anyway.
As I mentioned above, I made some improvements to solve the comment.

py9mrg commented:
Ah yeah sorry, I'm being stupid. I mentioned katex rendering being a solution for offline equation rendering fo `xaringan` too, I'm getting my wires crossed between the issues. Will post the link there instead.

jeroen commented:
A new package `katex` is now on CRAN, which enables server-side math-to-html rendering in R, such that no javascript library is needed in the html document: https://docs.ropensci.org/katex/reference/katex.html.

cderv commented:
Thanks @jeroen ! That is a really great solution. Offering support directly into **rmarkdown** `math:` YAML key seems something that would be interesting. I'll look into this.

cderv reviewed (commented):
it has been a long time - sorry for the delay @atusy ! We are focusing on **rmarkdown** currently so we should get this merge soon. 

Improving the math support is important and this is possible an impactful change so I want to go thoroughly through this as there are several options. 

I did review the current PR as-is but I have a few generic questions / topic in mind for this PR

Self contained now works better with Pandoc. Including for katex. However, it will create heavy document with longer rendering as it will download the JS lib and insert it in the document. We could then continue to provide some tricks to insert the required JS code. However, we could probably change it. 

Patching the template file would be a way to insert it instead of using a JS script to modify the header. We could detect where the `$math$` variable in the template is and replace with our content. This way we could support any template with self contained, as long as they have the if clause with the `$math$` variable.  BTW, this is what is done in Quarto for next generation of this. 

Regarding self-contained for KaTeX, we could also default to **katex** R package (PR #2215) for server side rendering during post processing step. Or just offer a way to opt-in. We need to take this other PR into account for this PR. 

Regading the YAML header, I wonder if we should not play with names but have something like this (also Quarto inspired): 
````
output: 
  html_document:
    math: katex
````

````
output: 
  html_document:
    math: 
       engine: katex
       url: https://cdn.jsdelivr.net/npm/katex@0.12.0/dist"
````

Or use exactly like in Pandoc's default file (which Quarto will use)
````
output: 
  html_document:
    html-math-method:
      method: mathjax
      url: "https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"
````

Doing this could help with transition if we do need to use conversion tool. We could also add cleanly some specific method like katex R package


I would think the latter to be convenient as this would be how to config with Pandoc directly. 

Last topic is the lifecycle of this feature. Mathjax support is here since a long time in **rmarkdown** and I think we need to think of deprecation. The good and clean practice IMO would be to deprecate `mathjax` argument properly. For internal function, I think it is also good practice to deprecate just in case it is used somewhere. We don't yet use [**lifecycle**](https://lifecycle.r-lib.org) R package for this, but it would be a good opportunity to start IMO. This would help to offer clean lifecycle information.

cderv commented:
@atusy Thanks again for this great PR ! Supporting all Maths engine is really important IMO. 
Following this review and a discussion with @yihui , is that ok if I take the lead on the PR now ? Does the below seems reasonable to you before I start ? 

What we would do: 

- [ ] Patch the template now to insert the math part when self_contained = TRUE. Meaning replacing `$math$` with some content that we have today in the HTML template. This would allow support custom templates too.
- [ ] Use YAML field to select the method. Are you ok with this names ? Using fixed names seems better to me, all the more because you can't select several math engine.

       ````
       output: 
         html_document:
           math: 
             engine: katex
             url: https://cdn.jsdelivr.net/npm/katex@0.12.0/dist"
      ````

       ````
       output: 
         html_document:
           math: mathml
      ````
- [ ] Add support for R **katex** as a specific engine. 

       ````
       output: 
         html_document:
           math: r-katex
      ````

- [ ] Don't bother with deprecation here. `mathjax` argument can stay and those internal function can be removed. We'll just need a companion **pagedown** PR. 

@yihui did I miss anything ? 

Thanks.

yihui commented:
That's a good summary of our discussion. Thanks!

atusy commented on R/html_document.R:
```diff
@@ -673,8 +683,13 @@ html_highlighters <- function() {
   c(highlighters(), "textmate")
 }
 
-default_mathjax <- function() {
-  paste0("https://mathjax.rstudio.com/latest/", mathjax_config())
```
Wow, I am surprised. Does CRAN allow usage of `:::`?
Anyway, I will keep the function.
I would recommend export the function to avoid such a spurious removal.

atusy commented on R/html_document_base.R:
```diff
@@ -19,6 +19,7 @@
 html_document_base <- function(theme = NULL,
                                self_contained = TRUE,
                                lib_dir = NULL,
+                               math = "default",
                                mathjax = "default",
```
got it

atusy commented on R/pandoc.R:
```diff
@@ -472,44 +472,77 @@ pandoc_self_contained_html <- function(input, output) {
+         "(you should set self_contained to FALSE)", call. = FALSE)
 }
 
-pandoc_mathjax_args <- function(mathjax,
-                                template,
-                                self_contained,
-                                files_dir,
-                                output_dir) {
```
In general, packages should not directly depend on internal functions from other packages...
Anyway, we can start from deprecating the function.

cderv replied:
For reference in the context of **rmarkdown**, we always need to check that the RStudio IDE code does not use one internal function. It can happen - RStudio IDE use some internal functions of package sometimes. (until they implement their own version)

cderv replied:
It seems like it is not used. So we'll probably remove this function.

atusy commented on R/html_document_base.R:
```diff
@@ -36,10 +37,13 @@ html_document_base <- function(theme = NULL,
   args <- c()
 
   # self contained document
+  if (identical(math, "default")) {
+    math <- c(mathjax = mathjax)
+  }
```
Good idea! I also want to deprecate.

cderv replied:
For now I adapted so that it is still supported. But I'll probably add a message for deprecation

atusy commented on R/pandoc.R:
```diff
@@ -472,44 +472,77 @@ pandoc_self_contained_html <- function(input, output) {
+    }
+  }
+
+  if (identical(engine, "webtex")) {
+    return(if (identical(url, "default")) {
+      "--webtex"
+    } else {
+      paste0("--webtex=", url)
+    })
+  }
```
This is because `mathml` and `gladtex` does not require URL like `--mathml=url`.
Yet, my implementation lacks the alignment...

https://pandoc.org/MANUAL.html#math-rendering-in-html

I think the following form which you suggest is the better.
In this way, we can simply ignore `url` for `mathml` and `gladtex`

```yaml
output: 
  html_document:
    math: 
       engine: katex
       url: https://cdn.jsdelivr.net/npm/katex@0.12.0/dist"
```

cderv replied:
This is the supported syntax now
````yaml
output: 
  html_document:
    math: 
       engine: katex
       url: https://cdn.jsdelivr.net/npm/katex@0.12.0/dist"
````

atusy commented:
@cderv
Thanks for the neat review! I was replying from the top to the bottom, and just realized you want to take the lead.
Yes, go ahead.
Unfortunately, I recently have less time to work on...

atusy commented on R/pandoc.R:
```diff
@@ -472,44 +472,77 @@ pandoc_self_contained_html <- function(input, output) {
-  if (!is.null(mathjax)) {
+  is_mathjax <- identical(engine, "mathjax")
+  is_katex <- identical(engine, "katex")
+
+  if (is_mathjax || is_katex) {
 
-    if (identical(mathjax, "default")) {
+    if (identical(url, "default")) {
       if (identical(template, "default"))
-        mathjax <- default_mathjax()
+        url <- default_math(engine)
       else
-        mathjax <- NULL
-    }
-    else if (identical(mathjax, "local")) {
+        url <- NULL
+    } else if (is_mathjax && identical(url, "local")) {
       mathjax_path <- pandoc_mathjax_local_path()
       mathjax_path <- render_supporting_files(mathjax_path,
                                               files_dir,
                                               "mathjax-local")
-      mathjax <- paste(normalized_relative_to(output_dir, mathjax_path), "/",
+      url <- paste(normalized_relative_to(output_dir, mathjax_path), "/",
                        mathjax_config(), sep = "")
     }
 
     if (identical(template, "default")) {
-      args <- c(args, "--mathjax")
-      args <- c(args, "--variable", paste0("mathjax-url:", mathjax))
+      args <- c(args, paste0("--", engine))
+      args <- c(args, "--variable", paste0(engine, "-url:", url))
     } else if (!self_contained) {
-      args <- c(args, paste(c("--mathjax", mathjax), collapse = "="))
+      args <- c(args, paste(c(paste0("--", engine), url), collapse = "="))
     } else {
-      warning("MathJax doesn't work with self_contained when not ",
+      warning("MathJax and KaTeX do not work with self_contained when not ",
               "using the rmarkdown \"default\" template.", call. = FALSE)
```
> Do we know if this is still relevant and true ?

Sorry no.

> default to server side rendering

I do not prefer it because users have to install katex package, which I guess will be a *suggested*, not *imported*, package in **rmarkdown**.

cderv replied:
If we default to server side rendering, we would make the package import. But maybe that is not worth it and it should be opt-in.

cderv replied:
I remove this warning as I don't think we should block usage of Math with non default template. Self contained should probably work with pandoc in recent version

cderv commented on inst/rmd/h/default.html:
```diff
@@ -479,6 +479,21 @@ <h2 id="$idprefix$toc-title">$toc-title$</h2>
+$endif$
+
+$if(katex-url)$
+<link rel="stylesheet" href="$katex-url$/katex.min.css" data-external="1">
```
However, I kept that. Math engine other than mathjax require Pandoc 2.0 now.

cderv commented on R/pandoc.R:
```diff
@@ -472,44 +472,77 @@ pandoc_self_contained_html <- function(input, output) {
-                                self_contained,
-                                files_dir,
-                                output_dir) {
+pandoc_math_args <- function(math,
```
I have refactored the logic 

- `pandoc_math_args()` is like other `pandoc_*_args()` function : a lower function to create a command line arg
- `add_math_support()` has now the main logic, including setting the default behavior. This is in the continuation of other feature regrouped by function like `add_anchor_sections()`

cderv commented:
I have integrated the first part of the change by refactoring the logic and enabling the YAML syntax described in https://github.com/rstudio/rmarkdown/pull/1940#issuecomment-1015597217

For now, I believe the change won't break anything as I believe I was cautious to insure backward compatibility

`mathjax` argument is still taken into account if provided and non `default`, and a `math` argument is build based on it. The template as not changed for Mathjax currently. 

Next step would be to patch template, meaning putting back
````
$if(math)$
$math$
$end$
````
in our template. I don't think other engine use it for now, but it would be better to keep them. 
We would be replacing this part of the template by the correct script part, either for MathJaX or for KaTeX. 

I don't know how doing that would affect other tooling (like **learnr** or shiny document) - this would require more tests I believe. 

I am currently testing several outputs after adding some unit tests.

cderv commented:
I have made sure here that `math` will only be used when `mathjax = "default"`. I need to check in extension format if this is a valide choice or not. I believe this would insure we don't break anything yet. We could take the opportunity to add a deprecation warning to use `math` argument instead of `mathjax` (using **lifecycle** package for example)  

<details>
<summary>Example of message we can get</summary>

``` r
lifecycle::deprecate_warn(
    "2.11.3",
    "rmarkdown::html_document_base(mathjax)", 
    "rmarkdown::html_document_base(math)")
#> Warning: The `mathjax` argument of `html_document_base()` is deprecated as of rmarkdown 2.11.3.
#> Please use the `math` argument instead.
#> This warning is displayed once every 8 hours.
#> Call `lifecycle::last_lifecycle_warnings()` to see where this warning was generated.
```


</details>

So I started to check support for other HTML format. I then wondered 'shouldn't we use a `htmlDependency()` for katex so that it can be easily included in any template ? 🤔 At least any of ours, instead of having to copy paste the HTML part in each template. So I am trying that. 

Here are some notes about math support in other outputs that I'll keep updating: 

- [ ] `html_fragment()` sets `--mathjax` when `mathjax = TRUE`, should we support other engine using `math` ? Otherwise backward compatibility

cderv commented on R/html_document.R:
```diff
@@ -612,7 +637,9 @@ html_document <- function(toc = FALSE,
                                      self_contained = self_contained,
-                                     lib_dir = lib_dir, mathjax = mathjax,
+                                     lib_dir = lib_dir,
+                                     math = math,
+                                     mathjax = mathjax,
```
I need to reopen the discussion about this option name. I was not expecting this but it is causing issue because of R partial argument matching in functions. 
This is creating issues with other format that extend this one : 

Example of behavior
```r
# base format defines a new argument math
base_format <- function(math = "katex", mathjax = "default"){ math } 
# expected
base_format(math = "dummy")
#> [1] "dummy"

# another function use this format not knowing explicitly the new argument
out_format <- function(mathjax = "default", ...) { base_format(mathjax = mathjax, ...)}
# Not expected - mathjax is matched because `...` is after `mathjax`
out_format(math = "dummy")
#> [1] "katex"
```

Changing the argument name would solve this - or it would require to explictly set the argument in all extension format. It seems not right to keep that IMO.

cderv replied:
Argument name was changed to `math_method`

cderv commented:
Last changes includes 

* Inserting KaTeX HTML in head using a html dependency.  This should allow us to bring support to any format as we don't have to modify the Pandoc's template. For example, this works ok now: 
```r
output:
  learnr::tutorial:
    math_method: katex
```
* For this to work, argument has been renamed to `math_method`
* Support has been brought to `slidy_presentation()` and `html_notebook()`

I'll see if I can took this opportunity to remove the MathJax part from the template also using a similar trick or do template patching.  

A few more tests with shiny runtime and our other formats, and this should be good to merge.

cderv commented:
@yihui I tested with more formats and it is working well. 

I would suggest the following: 

We could merge the current PR which brings support for new engines **without changing anything** regarding Mathjax. This means that KaTeX dependency will included only when `template = "default"` is passed to `html_document_base()`, as for mathajx, which generally support that a custom template is passed with the necessary content. 
This is how everything currently works (**learnr**, **flexdashboard**, ...) and I think we could improve this in a separate PR. 

So I would try to work on improving mathjax integration with a _patching_ process, and no template constraint, probably trying to remove this template constraint, but in another PR. 

Is this ok for you merge this one and build the rest upon it ? (R package **katex** support, github_document() math support, ...)

yihui reviewed (approved):
Yes, I think it's a better idea to make the improvements in separate PRs. Thanks!

atusy commented:
Thanks @cderv !
